stdlib/InteractiveUtils/src/macros.jl

```diff
@@ -9,8 +9,34 @@ using Base: insert!, replace_ref_begin_end!,
 # via. `Base.@time_imports` etc.
 import Base: @time_imports, @trace_compile, @trace_dispatch
 
-typesof_expr(args::Vector{Any}, where_params::Union{Nothing, Vector{Any}} = nothing) = rewrap_where(:(Tuple{$(Any[esc(reescape(get_typeof, a)) for a in args]...)}), where_params)
-typesof_expr_unescaped(args::Vector{Any}, where_params::Union{Nothing, Vector{Any}} = nothing) = rewrap_where(:(Tuple{$(Any[reescape(get_typeof, a) for a in args]...)}), where_params)
+typesof_expr(args::Vector{Any}, where_params::Union{Nothing, Vector{Any}} = nothing) = rewrap_where(:(Tuple{$(Any[esc(reescape(get_typeof, a)) for a in extract_args(args)]...)}), where_params)
+typesof_expr_unescaped(args::Vector{Any}, where_params::Union{Nothing, Vector{Any}} = nothing) = rewrap_where(:(Tuple{$(Any[reescape(get_typeof, a) for a in extract_args(args)]...)}), where_params)
+
+function extract_args(args)
+    stop = 0
+    for (i, arg) in enumerate(args)
+        arg = Meta.unescape(arg)
+        encountered_vararg = stop !== 0
+        isexpr(arg, :(::)) && encountered_vararg && error("Explicit type annotations cannot come after a `::Vararg{...} or `::T...` annotation")
+        is_vararg = isexpr(arg, :...) && isexpr(arg.args[1], :(::)) ||
+                    # technically, a user may define their own `Vararg` binding,
+                    # but handling this right would require deferring the check until runtime,
+                    # where we can't differentiate between a type coming from a value or a type annotation
```

Comment:
Can you explain more about this ambiguity?

`typeof` should never yield a (naked) `Vararg{...}` afaik

Reply:
Oh! To implement:
> if you have explicit type-ascriptions after the Vararg, that should arguably be an error

I see.

As someone who has created a `Vararg` type in a package before (it was a partial implementation of Julia's subtyping, so not too surprising it showed up), I would like to avoid making syntactic assumptions on what "Vararg" means

If it's necessary, I think I am slightly more in favor of ignoring any types after the Vararg (explicit or not), if it means restoring that

Reply:
I moved to check to runtime in ac19f4ee8e94951d2355a86aaadaffa50e3bace5, so we no longer have to rely on these fishy assumptions. Do you like it better now?